### PR TITLE
Merge experiments into main

### PR DESCRIPTION
**Pull request by bw4sz**
This PR brings my work into the main branch.
# Summary
* Remove loss weighting to improve empty frame stability
* Add photoshop annotations for weak training from Lindsey's data
* Scripts to generate weak annotations for mining
* A few tests to ensure empty annotations are working as intended
* An update to the empty frame csv
* Add an anhinga class for training purposes to reduce confusion

## Sample run 
https://www.comet.com/bw4sz/everglades-species/ea8eae1c86104e718a9376ab33d1c0e7?experiment-tab=chart&showOutliers=true&smoothing=0&transformY=smoothing&viewId=1WPBkIDqjQNaRp16py6x3TUUa&xAxis=wall

### CODE REVIEW

**Comment by bw4sz** [comment]
Note that this is supposed to be squashed before merge.

File: experiment.sh [hunk 1]
```diff
@@ -1,24 +1,17 @@
...
 #SBATCH --partition=gpu
-#SBATCH --gpus=a100:1
+#SBATCH --gpus=a100:3
+
 ulimit -c 0
-ml git
-cd /blue/ewhite/everglades/EvergladesSpeciesModel
-git checkout $1
-source activate ESM
+source activate DeepTreeAttention
```
Comment (ethanwhite) [inline experiment.sh:16]: This should be the environment created by the environment.yml in the repo, which currently creates an environment named ESM. If the environment needs updating it should be done by updating the environment.yml file to include the necessary dependencies.

File: create_empty_frame_csv.py [hunk 2]
```diff
@@ -0,0 +1,9 @@
+# Iterate through the empty folder and create a csv folder for model training/evaluation
+from glob import glob
+import pandas as pd
+import os
+
+images = glob("/blue/ewhite/everglades/empty-frames/**/*.png", recursive=True)
+#images = [os.path.basename(x) for x in images]
```
Comment (ethanwhite) [inline create_empty_frame_csv.py:7]: If we don't need this commented out line it should be dropped. If it's used periodically for some purpose that should be documented with a comment.

File: evaluate_only.py [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+    dirname = os.path.dirname(test_path)
+    test_frame_df["image_path"] = test_frame_df["image_path"].apply(lambda x: os.path.join(dirname,x))
+    empty_images = test_frame_df.image_path.unique()  
+    
+    model.config["score_thresh"] = 0.01
+    predict_empty_frames(model, empty_images, comet_logger, invert=True)
+    
+    #Test on empy frames
+    if empty_images_path:
+        model.config["score_thresh"] = 0.01
```
Comment (ethanwhite) [inline evaluate_only.py:145]: Add a brief commend between Lines 144 and 145 explaining the threshold switching.

File: everglades_species.py [hunk 4]
```diff
@@ -150,17 +85,35 @@ def train_model(train_path, test_path, empty_images_path=None, save_dir=".",
...
-
+    
+    #Add weak annotations from photoshop to train
+    weak_train = pd.read_csv("/blue/ewhite/everglades/photoshop_annotations/split_annotations.csv")
+    train = pd.concat([train, weak_train])
+    train = train[train.label.isin(['Great Egret', 'Roseate Spoonbill', 'White Ibis',
+           'Great Blue Heron', 'Wood Stork', 'Snowy Egret',"Anhinga"])]
+    test = test[test.label.isin(train.label)]
+    
+    #add in weak annotations
```
Comment (ethanwhite) [inline everglades_species.py:96]: Change to `#add in weak empty frame annotations`

File: everglades_species.py [hunk 5]
```diff
@@ -193,59 +146,29 @@ def train_model(train_path, test_path, empty_images_path=None, save_dir=".",
         comet_logger.experiment.log_parameters(model.config)
         comet_logger.experiment.log_parameter("Training_Annotations",train.shape[0])    
         comet_logger.experiment.log_parameter("Testing_Annotations",test.shape[0])
+        comet_logger.experiment.log_parameter("model_savedir",model_savedir)
         
-    im_callback = images_callback(csv_file=model.config["validation"]["csv_file"], root_dir=model.config["validation"]["root_dir"], savedir=model_savedir, n=20)    
-    model.create_trainer(callbacks=[im_callback], logger=comet_logger)
+    #im_callback = images_callback(csv_file=model.config["validation"]["csv_file"], root_dir=model.config["validation"]["root_dir"], savedir=model_savedir, n=20)
```
Comment (ethanwhite) [inline everglades_species.py:151]: Add a comment above this explaining when to uncomment it.

File: everglades_species.py [hunk 6]
```diff
@@ -288,7 +211,7 @@ def train_model(train_path, test_path, empty_images_path=None, save_dir=".",
             ytrue = torch.nn.functional.one_hot(ytrue.to(torch.int64), num_classes = model.num_classes + 1).numpy()
 
             # Add a label for undetected birds and create confusion matrix
-            model.label_dict.update({'Bird Not Detected': 6})
+            model.label_dict.update({'Bird Not Detected': 7})
```
Comment (ethanwhite) [inline everglades_species.py:214]: Can you explain this change to me so I understand it? We have 6 species, which I believe were encoded 0, 1, 2, 3, 4, 5, hence the addition with the 6. Is there a change to keep 0 for background somewhere else or something different?

**Review by ethanwhite — CHANGES_REQUESTED** [review]
Overall this looks great. I have a handful of detailed suggestions/questions.

My one bigger question is it looks like you've pulled a bunch of stuff out into `evaluate_only.py`, which makes sense, but I don't see that module getting used anywhere. So right now it seems like it's unlinked code for anyone but you. Am I missing something?

File: everglades_species.py (hunk 6 above)
Comment (bw4sz) [reply everglades_species.py:214]: Added Anhinga.

Comment (bw4sz) [reply everglades_species.py:214]: Just for model training, no test, kinda like a distraction class.

Comment (ethanwhite) [reply everglades_species.py:216]: This is a nice example of a case where compartmentalized commits can really help with communication. A single commit with this change that indicates that it is to add Anhinga would make it easy for someone who was confused to quickly understand the source of the `7` using blame (and make it easy for someone reviewing the PR to understand why the change was made).

File: empty_frames_utilities.py [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+            boxes = pd.DataFrame({"image_path":[path], "xmin":[None],"ymin":[None],"xmax":[None],"ymax":[None],"label":[None],"image":[path], "score":[None]})
+        precision_curve.append(boxes)
+    
+    precision_curve = pd.concat(precision_curve)
+    recall_plot = plot_recall_curve(precision_curve, invert=invert)
+    value = empty_image(precision_curve, threshold=0.3)
+    
+    if invert:
+        value = 1 - value
+        metric_name = "BirdRecall_at_0.3"
```
[the remark this answers is not in the document]
Comment (bw4sz) [reply empty_frames_utilities.py:42]: No, the naming here is the recall value at the desired threshold, I will rename.

File: everglades_species.py [hunk 8]
```diff
@@ -128,16 +64,15 @@ def index_to_example(index, results, test_path, comet_experiment):
...
 def train_model(train_path, test_path, empty_images_path=None, save_dir=".",
                 gbd_pretrain = True,
-                balance_classes = False, balance_min = 0, balance_max = 100000,
-                one_vs_all_sp = None,
                 experiment_name="ev-species",
                 debug = False):
     """Train a DeepForest model"""
     
-    comet_logger = CometLogger(project_name="everglades-species", workspace="weecology", experiment_name=experiment_name)
+    comet_logger = CometLogger(project_name="everglades-species", workspace="bw4sz", experiment_name=experiment_name)
```
[the remark this answers is not in the document]
Comment (bw4sz) [reply everglades_species.py:71]: In terms of separate users contributing the one dashboard, we can't do that unless we change comet tiers, or we get a joint user account. Good(ish) news is that the code can stay as /weecology/ so that it defaults to your account. My .comet.config 

```
(DeepTreeAttention) [b.weinstein@c0906a-s35 EvergladesSpeciesModel]$ cat ~/.comet.config
[comet]
api_key=###########
```
takes precedence. So it will get place in /bw4sz/

File: experiment.sh (hunk 1 above)
Comment (bw4sz) [reply experiment.sh:16]: Building now, very slow.

Comment (bw4sz) [reply experiment.sh:16]: Runs clean for me. Changed name back to ESM.

**Comment by bw4sz** [comment]
I believe all issues have been resolved.

File: experiment.sh [hunk 9]
```diff
@@ -1,24 +1,17 @@
...
```
Comment (ethanwhite) [inline experiment.sh:16]: I'm still seeing the change from `ESM` to `DeepTreeAttention` here

File: everglades_species.py [hunk 10]
```diff
@@ -128,16 +65,15 @@ def index_to_example(index, results, test_path, comet_experiment):
...
```
Comment (ethanwhite) [inline everglades_species.py:72]: Are we putting this back to `weecology` based on the discussion in Slack?

File: evaluate.py [hunk 11]
```diff
@@ -0,0 +1,185 @@
...
+        test_path: path to csv on disk, format image_path (relative to root dir), xmin, ymin, xmax, ymax, label
+        model_path: a deepforest saved model
+        empty_images_patH: path to csv on disk, same format as above, but with None, in xmin, ymin, xmax, ymax to denote empty frames
+        experiment_name: used to name a new comet_experiment, superceded by comet_logger
+        comet_logger: an existing comet logger to log metrics
+    Returns:
+        results: a pandas dataframe of deepforest results
+    """ 
+    if experiment_name is not None:
+        comet_logger = CometLogger(project_name="everglades-species", workspace="bw4sz", experiment_name=experiment_name)
```
Comment (ethanwhite) [inline evaluate.py:74]: Are we changing this to `weecology` based on the discussion in Slack?

File: everglades_species.py [hunk 12]
```diff
@@ -147,21 +83,39 @@ def train_model(train_path, test_path, empty_images_path=None, save_dir=".",
...
-        train["label"] = np.where(train["label"] == one_vs_all_sp, one_vs_all_sp, "Other Species")
-        test["label"] = np.where(test["label"] == one_vs_all_sp, one_vs_all_sp, "Other Species")
-
-    #Store test train split for run to allow multiple simultaneous run starts
+    
+    # Add weak annotations from photoshop to train
+    weak_train = pd.read_csv("/blue/ewhite/everglades/photoshop_annotations/split_annotations.csv")
+    train = pd.concat([train, weak_train])
+    train = train[train.label.isin(['Great Egret', 'Roseate Spoonbill', 'White Ibis',
+           'Great Blue Heron', 'Wood Stork', 'Snowy Egret',"Anhinga"])]
```
Comment (ethanwhite) [inline everglades_species.py:94]: Since there are a couple of other things to cleanup anyway - white space after the comma before Anhinga and single quotes for consistency with the rest of the list.

**Review by ethanwhite — CHANGES_REQUESTED** [review]
Just a couple of last little things. If I misunderstood your last comment on Slack then feel free to ignore the bw4sz -> weecology comments.

File: experiment.sh (hunk 9 above)
Comment (bw4sz) [reply experiment.sh:16]: We must be offset, pushing, I've got ESM here.

Comment (bw4sz) [reply experiment.sh:16]: I'm up to date, hard refresh?

Comment (bw4sz) [reply experiment.sh]: I think it fixed itself. 
image

Comment (ethanwhite) [reply experiment.sh]: :+1: looks good now
